### PR TITLE
Generic search changes.

### PR DESCRIPTION
**Pull request by 07Himank**
I worked on ... because ...

- centralized updating indexes code.
- made all scripts constants at one place
- remove deepCopy logic.
- remove search related code from entityRepository

#
### Type of change:

- [ x] Improvement

### CODE REVIEW

File: openmetadata-service/src/main/java/org/openmetadata/service/search/indexes/DashboardIndex.java [hunk 1]
```diff
@@ -76,6 +65,17 @@ public Map<String, Object> buildESDoc() {
     doc.put("entityType", Entity.DASHBOARD);
     doc.put("serviceType", dashboard.getServiceType());
     doc.put("fqnParts", suggest.stream().map(SearchSuggest::getInput).collect(Collectors.toList()));
+    if (dashboard.getOwner() != null) {
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/indexes/DashboardIndex.java:68]: @07Himank why do we need to do this? we are just capturing the owner entity reference as it is today

File: openmetadata-service/src/main/java/org/openmetadata/service/search/indexes/DashboardDataModelIndex.java [hunk 2]
```diff
@@ -57,7 +46,17 @@ public Map<String, Object> buildESDoc() {
...
             dashboardDataModel.getFullyQualifiedName(),
             suggest.stream().map(SearchSuggest::getInput).collect(Collectors.toList())));
-
+    if (dashboardDataModel.getOwner() != null) {
+      EntityReference owner = dashboardDataModel.getOwner();
+      owner.setDisplayName(CommonUtil.nullOrEmpty(owner.getDisplayName()) ? owner.getName() : owner.getDisplayName());
+      doc.put("owner", owner);
+    }
+    if (dashboardDataModel.getDomain() != null) {
+      EntityReference domain = dashboardDataModel.getDomain();
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/indexes/DashboardDataModelIndex.java:55]: all of this code can be centralized in the interface with default method

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java [hunk 3]
```diff
@@ -694,17 +717,46 @@ public final PutResponse<T> createOrUpdate(UriInfo uriInfo, T updated) {
 
   @SuppressWarnings("unused")
   protected void postCreate(T entity) {
-    if (supportsSearchIndex) {
-      String contextInfo = entity != null ? String.format("Entity Info : %s", entity) : null;
-      searchClient.updateSearchEntityCreated(JsonUtils.deepCopy(entity, entityClass));
+    if (supportsSearch) {
+      searchRepository.updateSearchEntityCreated(entity);
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java:721]: can we rename this method as searchReposiory.createEntity?

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+  default void handleOwnerUpdates(EntityInterface original, EntityInterface updated, String eventType) {
+    if (eventType.equalsIgnoreCase("added")) {
+      this.updateSearchChildrenUpdated(
+          updated,
+          getOwnerChangeScript(eventType, ""),
+          updated.getEntityReference().getType() + ".id",
+          updated.getId().toString(),
+          ENTITY_TO_CHILDREN_MAPPING.get(updated.getEntityReference().getType()).toString(),
+          updated.getOwner());
+    } else if (eventType.equalsIgnoreCase("updated")) {
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:78]: lets create final strings for added, updated, deleted

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+  String DEFAULT_UPDATE_SCRIPT = "for (k in params.keySet()) { ctx._source.put(k, params.get(k)) }";
+  String CLASSIFICATION_DISABLE_SCRIPT = "ctx._source.disabled=%s";
+  String REMOVE_DOMAINS_CHILDREN_SCRIPT = "ctx._source.remove('domain')";
+  String REMOVE_TAGS_CHILDREN_SCRIPT =
+      "for (int i = 0; i < ctx._source.tags.length; i++) { if (ctx._source.tags[i].tagFQN == '%s') { ctx._source.tags.remove(i) }}";
+  String REMOVE_TEST_SUITE_CHILDREN_SCRIPT =
+      "for (int i = 0; i < ctx._source.testSuites.length; i++) { if (ctx._source.testSuites[i].id == '%s') { ctx._source.testSuites.remove(i) }}";
+
+  default void handleOwnerUpdates(EntityInterface original, EntityInterface updated, String eventType) {
+    if (eventType.equalsIgnoreCase("added")) {
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:70]: can we turn this into switch case

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+  }
+
+  default void handleClassificationUpdate(Classification entity) {
+    this.updateSearchEntityUpdated(entity, String.format(CLASSIFICATION_DISABLE_SCRIPT, entity.getDisabled()), "");
+  }
+
+  default String getOwnerChangeScript(String eventType, String ownerId) {
+    String scriptTxt = "";
+    if (eventType.equals("added")) {
+      scriptTxt = "if(ctx._source.owner == null){ ctx._source.put('owner', params)}";
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:134]: can we declare these scripts as final strings?

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+      scriptTxt = String.format("if(ctx._source.domain.id == '%s'){ ctx._source.remove('domain')}", domainId);
+    } else if (eventType.equals("updated")) {
+      scriptTxt = String.format("if(ctx._source.domain.id == '%s'){ ctx._source.put('domain', params)}", domainId);
+    }
+    return scriptTxt;
+  }
+
+  default void handleEntityDeleted(EntityInterface entity) {
+    switch (entity.getEntityReference().getType()) {
+      case Entity.DOMAIN:
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:157]: why we need this switch case, can't we make this generic code? we don't want to keep adding to this whenever we add a new entity right?

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+
+  default void updateElasticSearch(UpdateRequest updateRequest) throws IOException {
+    throw new CustomExceptionMessage(Response.Status.NOT_IMPLEMENTED, NOT_IMPLEMENTED_METHOD);
+  }
+
+  default void updateElasticSearch(org.elasticsearch.action.update.UpdateRequest updateRequest) throws IOException {
+    throw new CustomExceptionMessage(Response.Status.NOT_IMPLEMENTED, NOT_IMPLEMENTED_METHOD);
+  }
+
+  void updateSearchEntityCreated(EntityInterface entity);
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:254]: call this as createEntity

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+
+  void updateSearchEntityCreated(EntityTimeSeriesInterface entity);
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:256]: createTimeSeriesEntity

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+
+  void deleteByScript(String index, String scriptTxt, HashMap<String, Object> params);
+
+  void updateSearchEntityDeleted(EntityInterface entity, String script, String field, String alias);
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:260]: deleteEntity

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+
+  void softDeleteOrRestoreEntityFromSearch(EntityInterface entity, boolean delete);
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:262]: softDeleteOrRestoreEntity

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+
+  void softDeleteOrRestoreChildrenFromSearch(EntityInterface entity, boolean delete, String field, String alias);
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:264]: softDeleteOrRestoreChildren

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+
+  void updateSearchEntityUpdated(EntityInterface entity, String script, String field);
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:266]: updateEntity

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 14]
```diff
@@ -0,0 +1,406 @@
...
+
+  void updateSearchChildrenUpdated(
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:268]: updateChildren

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 15]
```diff
@@ -0,0 +1,406 @@
...
+      EntityInterface entity, String scriptTxt, String field, String value, String alias, Object data);
+
+  void close();
+
+  default BulkResponse bulk(BulkRequest data, RequestOptions options) throws IOException {
+    throw new CustomExceptionMessage(Response.Status.NOT_IMPLEMENTED, NOT_IMPLEMENTED_METHOD);
+  }
+
+  default org.elasticsearch.action.bulk.BulkResponse bulk(
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:277]: we are defaulting to elasticsearch classes?

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java [hunk 16]
```diff
@@ -0,0 +1,406 @@
...
+  default String getIndexMapping(SearchIndexDefinition.ElasticSearchIndexType elasticSearchIndexType, String lang)
+      throws IOException {
+    InputStream in =
+        SearchIndexDefinition.class.getResourceAsStream(
+            String.format(elasticSearchIndexType.indexMappingFile, lang.toLowerCase()));
+    assert in != null;
+    return new String(in.readAllBytes());
+  }
+
+  CollectionDAO getDao();
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:373]: why are we declaring this here? what is the need
and lets move this at the top

File: openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchClientImpl.java [hunk 17]
```diff
@@ -1011,27 +1018,17 @@ public void updateSearchEntityDeleted(EntityInterface entity, String scriptTxt,
...
   @Override
-  public void deleteEntityAndRemoveRelationships(EntityInterface entity, String scriptTxt, String field) {
+  public void softDeleteOrRestoreEntityFromSearch(EntityInterface entity, boolean delete) {
     if (entity != null) {
       String entityType = entity.getEntityReference().getType();
       SearchIndexDefinition.ElasticSearchIndexType indexType = IndexUtil.getIndexMappingByEntityType(entityType);
-      DeleteRequest deleteRequest = new DeleteRequest(indexType.indexName, entity.getId().toString());
-      deleteRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+      UpdateRequest updateRequest = new UpdateRequest(indexType.indexName, entity.getId().toString());
+      String scriptTxt = "ctx._source.deleted=" + delete;
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchClientImpl.java:1026]: we should declare all the scriptTexts in one place and call this softDeleteOrRestoreScriptTxt

File: openmetadata-service/src/main/java/org/openmetadata/service/search/indexes/ContainerIndex.java [hunk 18]
```diff
@@ -77,6 +64,12 @@ public Map<String, Object> buildESDoc() {
         getFQNParts(
             container.getFullyQualifiedName(),
             suggest.stream().map(SearchSuggest::getInput).collect(Collectors.toList())));
+    if (container.getOwner() != null) {
+      doc.put("owner", getOwnerWithDisplayName(container.getOwner()));
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/indexes/ContainerIndex.java:68]: may be we should index nulls as well, then we can support if the docs do not have owner or not

File: openmetadata-service/src/main/resources/elasticsearch/jp/storage_service_index_mapping.json [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+          "filter" : [
+            "kuromoji_baseform",
+            "kuromoji_part_of_speech",
+            "kuromoji_number",
+            "kuromoji_stemmer"
+          ]
+        },
+        "om_ngram": {
+          "tokenizer": "ngram",
+          "min_gram": 1,
```
Comment (harshach) [inline openmetadata-service/src/main/resources/elasticsearch/jp/storage_service_index_mapping.json:26]: set min_gram to 3 and max_gram to 10

File: openmetadata-service/src/main/resources/elasticsearch/jp/storage_service_index_mapping.json [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+      "serviceType": {
+        "type": "keyword"
+      },
+      "entityType": {
+        "type": "keyword"
+      },
+      "description": {
+        "type": "text",
+        "analyzer": "om_analyzer_jp",
+        "fields": {
```
Comment (harshach) [inline openmetadata-service/src/main/resources/elasticsearch/jp/storage_service_index_mapping.json:77]: remove all the ngrams on description. This is getting lot false positives in search results as it matches every single character in description

File: openmetadata-service/src/main/resources/elasticsearch/en/storage_service_index_mapping.json [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+        }
+      },
+      "serviceType": {
+        "type": "keyword"
+      },
+      "description": {
+        "type": "text",
+        "analyzer": "om_analyzer",
+        "fields": {
+          "ngram": {
```
Comment (harshach) [inline openmetadata-service/src/main/resources/elasticsearch/en/storage_service_index_mapping.json:72]: remove ngram for description in en language

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java (hunk 15 above)
Comment (07Himank) [reply openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java:277]: we have both declared .. for some reason we did this way .. exactly not recollecting

File: openmetadata-service/src/main/java/org/openmetadata/service/search/indexes/ContainerIndex.java (hunk 18 above)
Comment (07Himank) [reply openmetadata-service/src/main/java/org/openmetadata/service/search/indexes/ContainerIndex.java:68]: Will handle this in other PR need some discussion on this with @karanh37 for UI inputs had a call with him.

File: ingestion/examples/sample_data/datasets/tables.json [hunk 22]
```diff
@@ -678,7 +678,7 @@
           },
           {
               "id": "1cda9ecb-f4c6-4ed4-8506-abe965b64c87",
-              "name": "0005_dim_address_clean",
+              "name": "dim_address_clean",
```
Comment (ayush-shah) [inline ingestion/examples/sample_data/datasets/tables.json:681]: @harshach reverted it to 0005_dim_address_clean as it was giving Validation error in sample data for lifecycle

**Pull request merged**